sklearn/decomposition/_nmf.py

```diff
@@ -803,23 +892,31 @@ def _fit_multiplicative_update(X, W, H, beta_loss='frobenius',
     previous_error = error_at_init
 
     H_sum, HHt, XHt = None, None, None
-    for n_iter in range(1, max_iter + 1):
+
+    if batch_size is None:
+        batch_size = n_samples
+
+    batches = gen_batches(n_samples, batch_size)
+    batches = itertools.cycle(batches)
+    n_batches = n_samples // batch_size + 1
+    n_steps = max_iter * n_batches
+    for n_i, batch in zip(range(n_steps), batches):
         # update W
         # H_sum, HHt and XHt are saved and reused if not update_H
         delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
-            X, W, H, beta_loss, l1_reg_W, l2_reg_W, gamma,
-            H_sum, HHt, XHt, update_H)
-        W *= delta_W
-
+            X[batch], W[batch], H, beta_loss, l1_reg_W, l2_reg_W,
+            gamma, H_sum, HHt, XHt, update_H)
```

Comment:
At transform time (update_H = 0), XHt is computed in the first call to `_multiplicative_update_w` and equals `X[batch] @ H.T` and then it is reused in the next iterations, while it should be recomputed for each minibatch.

We should compute XHt once and call `_multiplicative_update_w(..., XHt[batch], ...)`.

Reply:
Thanks @jeremiedbb !!! H_sum, HHt, and XHt should all be reinitialized at each minibatch, right?

Reply:
> Thanks @jeremiedbb !!! H_sum, HHt, and XHt should all be reinitialized at each minibatch, right?

Sorry, me again... H_sum and HHt do not contain X[batch]... understood...

Reply:
H_sum and HHt are fine, they do not depend on the batch. XHt is ``X @ H.T``. Since at transform we don't update H, it only needs to be computed once. Then for each batch we can pick the part we need, i.e. XHt[batch].

Roughly something like the following:
```py
XHt = X @ H.T
H_sum = ...
HHt = ...
for i in range(n_steps):
    delta_W = _multiplicative_update_w(X[batch], W[batch], ..., XHt[batch], ...)
    W[batch] *= delta_W

    if update_H:
        H, A, B = _multiplicative_update_H(X[batch], W[batch], ...)

        XHt = X @ H.T
        H_sum = ...
        HHt = ...
```